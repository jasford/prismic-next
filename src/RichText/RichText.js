/* eslint react/no-danger: 0, jsx-a11y/alt-text: 0 */
import Link from 'next/link';
import PropTypes from 'prop-types';
import React from 'react';
import PrismicRichText, { Elements } from 'prismic-richtext';

const RichText = ({ data, linkResolver }) => {
  if (!data) return null;

  let keyIndex = 0;

  const getProps = (element) => {
    const props = { key: keyIndex };
    if (element.label) props.className = element.label;
    return props;
  };

  const serializeEmbed = element => (
    <div
      data-oembed={element.oembed.embed_url}
      data-oembed-type={element.oembed.type}
      data-oembed-provider={element.oembed.provider_name}
      {...getProps(element)}
      dangerouslySetInnerHTML={{ __html: element.oembed.html }}
    />
  );

  const serializeHyperlink = (element, children) => {
    const link = linkResolver(element.data);
    const p = getProps(element);
    if (link.href && link.href.substr(0, 4) === 'http') {
      return <a href={link.href} target="_blank" {...p}>{children}</a>;
    }
    const resolved = linkResolver(element.data);
    if (resolved.onClick) {
      return <a role="button" {...resolved} {...p}>{children}</a>;
    }
    return (
      <Link {...resolved} {...p}>
        <a>{children}</a>
      </Link>
    );
  };

  const serializeImage = (element) => {
    const imgProps = { src: element.url };
    if (element.alt) imgProps.alt = element.alt;
    if (element.copyright) imgProps.copyright = element.copyright;
    const main = <img {...imgProps} {...getProps(element)} />;
    if (element.linkTo) {
      return serializeHyperlink({ data: element.linkTo }, main);
    }
    return main;
  };

  const serializeSpan = (content) => {
    const contents = content.split('\n');
    if (contents.length < 2) return content;
    const output = [];
    contents.forEach((el) => {
      keyIndex += 1;
      output.push(<span key={keyIndex}>{el}</span>);
      keyIndex += 1;
      output.push(<br key={keyIndex} />);
    });
    return output.slice(0, -1);
  };

  const serializeLabel = (element, children) => (
    <span {...getProps(element.data)}>{children}</span>
  );

  function serialize(type, element, content, children) {
    keyIndex += 1;
    const standardEl = {
      [Elements.heading1]: 'h1',
      [Elements.heading2]: 'h2',
      [Elements.heading3]: 'h3',
      [Elements.heading4]: 'h4',
      [Elements.heading5]: 'h5',
      [Elements.heading6]: 'h6',
      [Elements.paragraph]: 'p',
      [Elements.preformatted]: 'pre',
      [Elements.strong]: 'strong',
      [Elements.em]: 'em',
      [Elements.listItem]: 'li',
      [Elements.oListItem]: 'li',
      [Elements.list]: 'ul',
      [Elements.oList]: 'ol',
    }[type];
    if (standardEl) {
      return React.createElement(standardEl, getProps(element), children);
    }
    switch (type) {
      case Elements.image: return serializeImage(element);
      case Elements.embed: return serializeEmbed(element);
      case Elements.hyperlink: return serializeHyperlink(element, children);
      case Elements.label: return serializeLabel(element, children);
      case Elements.span: return serializeSpan(content);
      default: return '';
    }
  }

  const serialized = PrismicRichText.serialize(data, serialize);
  return (
    <React.Fragment>
      <style jsx>{`
        div :global(img) {
          max-width: 100%;
        }
      `}
      </style>
      {serialized}
    </React.Fragment>
  );
};

RichText.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string,
  })),
  linkResolver: PropTypes.func.isRequired,
};

RichText.defaultProps = {
  data: null,
};

export default RichText;
