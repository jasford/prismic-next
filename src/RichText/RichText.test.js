import React from 'react';
import renderer from 'react-test-renderer';

import RichText from './RichText';

// Sample linkResolver function for use in tests.
function linkResolver(data) {
  if (data.link_type === 'Web') {
    return { href: data.url };
  }
  if (data.type === 'page') {
    return {
      href: `/content?slug=${data.data.slug}&folder=${data.data.folder}`,
      as: `/content/${data.data.slug}`,
    };
  }
  if (data.type === 'stub') {
    return { onClick: () => null };
  }
  return { href: '/' };
}

describe('ui/RichText', () => {
  test('Renders simple standard tag content', () => {
    const content = [
      {
        type: 'heading1',
        text: 'This is an h1.',
        spans: [],
      },
      {
        type: 'heading2',
        text: 'This is an h2.',
        spans: [],
      },
      {
        type: 'heading3',
        text: 'This is an h3.',
        spans: [],
      },
      {
        type: 'heading4',
        text: 'This is an h4.',
        spans: [],
      },
      {
        type: 'heading5',
        text: 'This is an h5.',
        spans: [],
      },
      {
        type: 'heading6',
        text: 'This is an h6.',
        spans: [],
      },
      {
        type: 'paragraph',
        text: 'This is a paragraph',
        spans: [],
      },
      {
        type: 'preformatted',
        text: 'This is a preformatted text',
        spans: [],
      },
      {
        type: 'strong',
        text: 'This is a bold (strong) text',
        spans: [],
      },
      {
        type: 'em',
        text: 'This is a emphasized (em) text',
        spans: [],
      },
      {
        type: 'list-item',
        text: 'This is an un-ordered list item.',
        spans: [],
      },
      {
        type: 'list-item',
        text: 'This is another un-ordered list item.',
        spans: [],
      },
      {
        type: 'o-list-item',
        text: 'This is an ordered list item.',
        spans: [],
      },
      {
        type: 'o-list-item',
        text: 'This is another ordered list item.',
        spans: [],
      },
    ];
    const tree = renderer.create((
      <RichText linkResolver={linkResolver} data={content} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Renders spans', () => {
    const content = [
      {
        type: 'heading1',
        text: 'This is a bold h1.',
        spans: [{
          start: 10,
          end: 14,
          type: 'strong',
        }],
      },
      {
        type: 'heading2',
        text: 'This is an paragraph features emphasized text.',
        spans: [{
          start: 30,
          end: 40,
          type: 'em',
        }],
      },
    ];
    const tree = renderer.create((
      <RichText linkResolver={linkResolver} data={content} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Renders more complex content blocks', () => {
    const content = [
      {
        type: 'paragraph',
        text: 'This is a test of the content system, using all the ' +
          'features of Prismic rich text.\nThis line is not a new ' +
          'paragraph. It should just have a line break before it.',
        spans: [
          {
            start: 8,
            end: 14,
            type: 'strong',
          }, {
            start: 10,
            end: 36,
            type: 'em',
          }, {
            start: 64,
            end: 71,
            type: 'hyperlink',
            data: {
              link_type: 'Web',
              url: 'https://www.prismic.com',
            },
          },
        ],
      },
      {
        type: 'paragraph',
        text: 'This block has a link with an onClick event handler.',
        spans: [
          {
            start: 17,
            end: 21,
            type: 'hyperlink',
            data: {
              link_type: 'Document',
              type: 'stub',
            },
          },
        ],
      },
      {
        type: 'paragraph',
        text: 'This entire line has a label',
        spans: [],
        label: 'awesome',
      },
      {
        type: 'paragraph',
        text: 'Just this one word has a label.',
        spans: [{
          start: 14,
          end: 18,
          type: 'label',
          data: {
            label: 'super-hot',
          },
        }],
      },
      {
        type: 'image',
        url: 'https://webapp-starter.cdn.prismic.io/webapp-starter/f68a6e7' +
          'b08a5192cfacf0edc4f12d3f8d51dacf0_desktop.png',
        alt: null,
        copyright: null,
        dimensions: {
          width: 2048,
          height: 2600,
        },
        linkTo: {
          id: 'WdJRuCwAAFdPj0la',
          data: {
            slug: 'getting-started',
            folder: 'content',
          },
          type: 'page',
          tags: [],
          slug: 'getting-started',
          lang: 'en-us',
          link_type: 'Document',
          isBroken: false,
        },
      },
      {
        type: 'paragraph',
        text: 'The image above has a link. The one below does not:',
        spans: [],
      },
      {
        type: 'image',
        url: 'https://webapp-starter.cdn.prismic.io/webapp-starter/a078e201' +
          '71f6125a25b37ae80dcfe3e772639142_business-card.png',
        alt: 'Business card.',
        copyright: 'Super top secret.',
        dimensions: {
          width: 512,
          height: 404,
        },
      },
      {
        type: 'embed',
        oembed: {
          type: 'video',
          embed_url: 'https://www.youtube.com/watch?v=ferZnZ0_rSM',
          title: 'Anderson .Paak & The Free Nationals: NPR Music Tiny Desk ' +
            'Concert',
          provider_name: 'YouTube',
          thumbnail_url: 'https://i.ytimg.com/vi/ferZnZ0_rSM/hqdefault.jpg',
          author_url: 'https://www.youtube.com/user/nprmusic',
          thumbnail_width: 480,
          author_name: 'NPR Music',
          height: 270,
          width: 480,
          version: '1.0',
          html: '<iframe width="480" height="270" src="https://www.youtube' +
            '.com/embed/ferZnZ0_rSM?feature=oembed" frameborder="0" allowf' +
            'ullscreen></iframe>',
          provider_url: 'https://www.youtube.com/',
          thumbnail_height: 360,
        },
      },
    ];
    const tree = renderer.create((
      <RichText linkResolver={linkResolver} data={content} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Ignores things without valid data.', () => {
    const content = [
      {
        type: 'bogus',
        text: 'This is some bogus content that should be ignored.',
        spans: [],
      },
    ];
    const tree = renderer.create((
      <RichText linkResolver={linkResolver} data={content} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Renders null when not passed any data', () => {
    const tree = renderer.create((
      <RichText linkResolver={linkResolver} data={undefined} />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
