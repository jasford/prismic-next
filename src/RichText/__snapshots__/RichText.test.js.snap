// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/RichText Ignores things without valid data. 1`] = `
<div
  className="jsx-646358086 jsx-646358086"
>
  
</div>
`;

exports[`ui/RichText Renders more complex content blocks 1`] = `
<div
  className="jsx-646358086 jsx-646358086"
>
  <p>
    This is 
    <strong>
      a 
      <em>
        test
      </em>
    </strong>
    <em>
       of the content system
    </em>
    , using all the features of 
    <a
      href="https://www.prismic.com"
      target="_blank"
    >
      Prismic
    </a>
    <span>
       rich text.
    </span>
    <br />
    <span>
      This line is not a new paragraph. It should just have a line break before it.
    </span>
  </p>
  <p>
    This block has a 
    <a
      onClick={[Function]}
      role="button"
    >
      link
    </a>
     with an onClick event handler.
  </p>
  <p
    className="awesome"
  >
    This entire line has a label
  </p>
  <p>
    Just this one 
    <span
      className="super-hot"
    >
      word
    </span>
     has a label.
  </p>
  <a
    href="/content/getting-started"
    onClick={[Function]}
  >
    <img
      src="https://webapp-starter.cdn.prismic.io/webapp-starter/f68a6e7b08a5192cfacf0edc4f12d3f8d51dacf0_desktop.png"
    />
  </a>
  <p>
    The image above has a link. The one below does not:
  </p>
  <img
    alt="Business card."
    copyright="Super top secret."
    src="https://webapp-starter.cdn.prismic.io/webapp-starter/a078e20171f6125a25b37ae80dcfe3e772639142_business-card.png"
  />
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<iframe width=\\"480\\" height=\\"270\\" src=\\"https://www.youtube.com/embed/ferZnZ0_rSM?feature=oembed\\" frameborder=\\"0\\" allowfullscreen></iframe>",
      }
    }
    data-oembed="https://www.youtube.com/watch?v=ferZnZ0_rSM"
    data-oembed-provider="YouTube"
    data-oembed-type="video"
  />
</div>
`;

exports[`ui/RichText Renders null when not passed any data 1`] = `null`;

exports[`ui/RichText Renders simple standard tag content 1`] = `
<div
  className="jsx-646358086 jsx-646358086"
>
  <h1>
    This is an h1.
  </h1>
  <h2>
    This is an h2.
  </h2>
  <h3>
    This is an h3.
  </h3>
  <h4>
    This is an h4.
  </h4>
  <h5>
    This is an h5.
  </h5>
  <h6>
    This is an h6.
  </h6>
  <p>
    This is a paragraph
  </p>
  <pre>
    This is a preformatted text
  </pre>
  <strong>
    This is a bold (strong) text
  </strong>
  <em>
    This is a emphasized (em) text
  </em>
  <ul>
    <li>
      This is an un-ordered list item.
    </li>
    <li>
      This is another un-ordered list item.
    </li>
  </ul>
  <ol>
    <li>
      This is an ordered list item.
    </li>
    <li>
      This is another ordered list item.
    </li>
  </ol>
</div>
`;

exports[`ui/RichText Renders spans 1`] = `
<div
  className="jsx-646358086 jsx-646358086"
>
  <h1>
    This is a 
    <strong>
      bold
    </strong>
     h1.
  </h1>
  <h2>
    This is an paragraph features 
    <em>
      emphasized
    </em>
     text.
  </h2>
</div>
`;
